streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -17,98 +17,177 @@
+
+        // TODO: this should be moved to task when it transits to running from created / restoring
+        // then even if there's a long time between the initialization and the first txn that is also fine.
+        initialize();
     }
 
+    /**
+     * @throws StreamsException fatal error that should cause the thread to die
+     */
     @Override
-    public void init(final Producer<byte[], byte[]> producer) {
-        this.producer = producer;
+    public void initialize() {
+        maybeInitTxns();
     }
 
-    @Override
-    public <K, V> void send(final String topic,
-                            final K key,
-                            final V value,
-                            final Headers headers,
-                            final Long timestamp,
-                            final Serializer<K> keySerializer,
-                            final Serializer<V> valueSerializer,
-                            final StreamPartitioner<? super K, ? super V> partitioner) {
-        Integer partition = null;
+    private void maybeInitTxns() {
+        if (eosEnabled) {
+            // initialize transactions if eos is turned on, which will block if the previous transaction has not
+            // completed yet; do not start the first transaction until the topology has been initialized later
+            try {
+                producer.initTransactions();
+            } catch (final TimeoutException exception) {
+                final String errorMessage = "Timeout exception caught when initializing transactions for task " + taskId + ". " +
+                    "\nThe broker is either slow or in bad state (like not having enough replicas) in responding the request, " +
+                    "or the connection to broker was interrupted sending the request or receiving the response. " +
+                    "\n Consider overwriting `max.block.ms` to a larger value to avoid timeout errors";
 
-        if (partitioner != null) {
-            final List<PartitionInfo> partitions = producer.partitionsFor(topic);
-            if (partitions.size() > 0) {
-                partition = partitioner.partition(topic, key, value, partitions.size());
-            } else {
-                throw new StreamsException("Could not get partition information for topic '" + topic + "'." +
-                    " This can happen if the topic does not exist.");
+                // TODO: we do NOT try to handle timeout exception here but throw it as a fatal error
+                throw new StreamsException(errorMessage, exception);
+            } catch (final KafkaException exception) {
+                throw new StreamsException("Error encountered while initializing transactions for task " + taskId, exception);
+            }
+        }
+    }
+
+    private void maybeBeginTxn() {
+        if (eosEnabled && !transactionInFlight) {
+            try {
+                producer.beginTransaction();
+            } catch (final ProducerFencedException error) {
```

Comment:
I checked the source code and confirmed that `producer.beginTransaction();` would not throw UnknownProducerId, but only ProducerFencedException. More generally, here's my current conclusion:

1. producer.send can throw an `KafkaException` which wraps a `UnknownProducerId` or `ProducerFencedException`;

2. producer.send(callback), the callback could contain `UnknownProducerId` and other retriable exception.

3. producer.xxxTxn can throw `ProducerFencedException`.

4. producer.close can throw any fatal `KafkaException`.

5. producer.flush would not throw exception itself, but it would cause the send callback to complete with exception as in 2).

Reply:
Thanks for clarifying this... Maybe we should update the Producer docs, since this is enormously subtle, but also important for handling correctly.

Reply:
Sounds good.